### ISSUE TITLE
[MD] Bad timestamp of messages after reconnecting

### ISSUE THREAD

**Issue by burstfreeze** [adiwajshing/Baileys#868]
If I connect after having the bot offline for a while it will get all the messages. However the latest messages of the batch carry a wrong timestamp with them. Instead of the actual sent timestamp it shows the timestamp of when the connection was made.

Is this something that is expected to be in chats.update or somewhere? 

This is how the message looks, the bad time stamp is the second of connecting whereas the message was sent 4 hours earlier

```
{
  "messages": [
    {
      "key": {
 ..
      },
      "message": {
        "conversation": " "
      },
      "messageTimestamp": "1637000770",
      "status": "SERVER_ACK",
      "participant": " ",
      "pushName": " "
    }
  ],
  "type": "notify"
}
[
  {
    id: ' ',
    conversationTimestamp: 1637000770
  }
]
[
  {
    id: ' ',
    notify: ' '
  }
]
```

I am going through the logs right now. There are a few messages mentioning that message ID with other timestamps (cannot make sense of these) AND THEN a classic "No sender found .. failure in decrypting message".

It then did
```
Closing open session in favor of incoming prekey bundle
Closing session: SessionEntry {
```

and then 

```
{"level":30,"time": ,"pid": ,"hostname":" ","msgAttrs":{"from":" ","type":"text","id":" ","participant":" ","offline":"0","t":"1636984571","count":"6","notify":" "},"retryCount":1,"msg":"sent retry receipt"}
```

WHICH IS THE EXACT TIMESTAMP OF WHEN THE MESSAGE WAS SENT. Later we get

```
"messageStubType": "CIPHERTEXT",
      "messageStubParameters": [
        "No sender key for:
```

and eventually the actual message posted above. So basically we DO get the actual timestamp of the message once I guess. Now the question is whether it is expected behavior of the server to send the current timestamp instead of the sent timestamp OR if the client is expected to save the result from the message that carries the right timestamp in it.

**Comment by hardianto-cpu** [comment]
just like me when I want to get the messagetype from the `message` property but the message property is not found and instead `messageStupType` appears

**Comment by adiwajshing** [comment]
Cannot reproduce this, I'm using the timestamp sent to me by the server. I believe there is something else going on with your own code.

**Comment by burstfreeze** [comment]
_Sorry for the long post but during writing a found a fix for this that we can add to Baileys_

@adiwajshing I've verified it with a clean fork (except for commenting out the automatic response in example.ts). If you want I can also take a video of that behavior but what I quoted above is the pure Baileys response. I think the client is expected to cache the timestamp of the retry request because the timestamp the server sends is the timestamp of when it "re-sends" the message which technically is a valid timestamp for the protocol but not for the user interface. 


### Just now I have been able to address this in Baileys (tldr at the bottom)

I just tried and verified something **and I can definitely solve the issue for you** but I'm not good at JavaScript.

Below the `const msgRetryMap` I added
`var msgRetryTimestamps = {}`

In the `sendRetryRequest` before the `sendNode()` gets triggered I have added the following:

```
try {
                console.log(node.attrs)
                msgRetryTimestamps[node.attrs.id] = node.attrs.t
                console.log("TStamp: added ", node.attrs?.t, " to message ID ", node.attrs.id)
            } catch (e) {
                console.log("TStamp: could not cast", node)
            }
```

In the `ws.on('CB:message'` I have added

```
        console.log("EMITTING FROM CB:Message")
        console.log(msgRetryTimestamps)
```


Someone has sent me a message while I had this instance offline. I have cached the log if you need ALL messages but all in all:

After launching the session and receiving the message it threw `No sender key for` as it does. It goes into my print with `TStamp: added  1637062849  to message ID  3A..6`

Then, after it was able to decrypt later I can see the messages.upsert triggered in the `CB:message` block with the content of `messageStubParameters.. No sender key for`.

Then, probably after retrying, `CB:Message` causes another messages.upsert with the actual decrypted message content _but the bad timestamp_ (`"messageTimestamp": "1637064752"`). Now, where my fix comes in:

As you remember I put a print statement for my timestamp cache dictionary so before the decrypted message it prints my dictionary with
```
EMITTING FROM CB:Message
{
  '3A..6': '1637062849',
```

which is the exact message ID with the RIGHT timestamp.

In other words, all that is needed to resolve this is a dictionary caching the timestamp mentioned in the node, and then checking for this entry in the loop that goes through the dec.successes, reassigning the timestamp so the event prints the correct timestamp :)

### tldr

As said in my opening comment, the node for the retryReceipt carries the correct timestamp. If we declare a `var msgRetryTimestamps = {}` and then do `msgRetryTimestamps[node.attrs.id] = node.attrs.t` after receiving the retry information, we can check for these entries in the `fullMessages` and replace with the cached timestamp, I am extremely happy now :)

I can do a PR with dictionaries but you probably do everything with your own declared types maps etc. 

If you want me to do this the "stupid" way let me know and I am happy to do that and PR, and if you want to do it your own way I am happy to provide you all the information needed.

**Comment by adiwajshing** [comment]
Oh I see what you mean, this issue is on messages that are retried, correct?

**Comment by burstfreeze** [comment]
Hey adiwajshing that is correct, I did not notice that before but now with a recent batch I only had it on 4 messages in the same group, all of which were retried.

Saying it's the last messages of the batch was a false assumption because I only went for the (faulty) timestamp that wrongly pushed them to the most recent date 😅 So it can be any message if it has been retried

**Comment by adiwajshing** [comment]
Right, so I would recommend:
1. you keep track of the original timestamp of the messages failed -- you can get this because the failed message event will be fired
2. When the "fixed" message is received, it'll have the same ID as the failed message. You should correlate the two and prevent the timestamp from being replaced here

**Comment by burstfreeze** [comment]
Hey @adiwajshing sorry I am not sure if I understood your statement - do you want us to handle this client-side or do you want me to PR? 

**If client side**, I'm not aware of any event that Baileys emits for the first incomplete message with the bad timestamp. The second decrypted message with the bad timestamp appears in messages.upsert but I think the failed message does not get emitted anywhere, correct me if I am wrong

**If Baileys side**, sure I would love to PR. If you allow me to do it the hacky dictionary way as presented above you will have a PR in less than an hour 😅

**Comment by adiwajshing** [comment]
I would recommend this is solved on your private end and not on Baileys, because some use cases might require using the new retried timestamp

**Comment by burstfreeze** [comment]
@adiwajshing I do not see how anyone would want the wrong timestamp but it's ok if we do it client-side. However then what's left is section 1; in what way Baileys should emit the first event since at the moment unless I am missing something there is no event being triggered for the real timestamp.

Can I PR a `messages.failure` event? Then your logic stays exempt from that but people who need the real timestamp can listen to that event in example.ts

**Comment by burstfreeze** [comment]
@adiwajshing please check if https://github.com/burstfreeze/Baileys/compare/d523bf8981324634e18f19667f09c923766515b8...multi-device is okay for you, then I PR.

The naming would fit to `MessageInfoUpdate` but the `jid` parameter would be useless in that case and since you do not seem to use it anywhere in your code I'm not sure if it will stay.

This is why I defined `MessageInitialDateInfo` for the `messages.retry` event.

Flow is as described before, except now it's emitting an actual event:

- Messages throws `No keys for`
- my new event emits with: correct timestamp from 2 days ago
- old event emits with:
1) senderTimestamp from 2 hours ago
2) messageTimestamp from the 0 seconds ago (time of connecting)

Still don't know who'd need the wrong ones but anyway, please let me know if you have any objections to the addition.
